llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -3141,6 +3141,73 @@ static bool isNonEqualSelect(const Value *V1, const Value *V2, unsigned Depth,
          isKnownNonEqual(SI1->getFalseValue(), V2, Depth + 1, Q);
 }
 
+static bool isNonEqualPointersWithRecursiveGEP(const Value *A, const Value *B,
+                                               unsigned Depth,
+                                               const SimplifyQuery &Q) {
+  // Where A is a recursive GEP for an incoming value of PHI indicating a
+  // loop. B can be a ptr/GEP.
+  // If the PHI has 2 incoming values, one of it being the recursive GEP A
+  // and other a ptr at same base and at an same/higher offset than B we are
+  // only incrementing the pointer further in loop if offset of recursive GEP is
+  // greater than 0.
+  if (!A->getType()->isPointerTy() || !B->getType()->isPointerTy())
+    return false;
+
+  auto *GEPA = dyn_cast<GEPOperator>(A);
+  // Make sure GEPA pointer operand is a PHI.
+  if (!GEPA || !isa<PHINode>(GEPA->getPointerOperand()))
+    return false;
+
+  // Handle 2 incoming PHI values with one being a recursive GEP.
+  auto *PN = dyn_cast<PHINode>(GEPA->getPointerOperand());
+  if (PN->getNumIncomingValues() != 2)
+    return false;
+
+  // Recursive GEP in second incoming value. Always keep Recursive GEP as
+  // FirstInst
+  Value *FirstInst = nullptr, *SecondInst = nullptr;
+  for (unsigned i = 0; i != 2; ++i) {
+    // Check if A is a Recursive GEP in one of the incoming values.
+    if (PN->getIncomingValue(i) == A && GEPA->getNumIndices() == 1 &&
+        isa<Constant>(GEPA->idx_begin())) {
+      FirstInst = PN->getIncomingValue(i);
+      SecondInst = PN->getIncomingValue(1 - i);
+      continue;
+    }
+  }
+  if (FirstInst == nullptr || SecondInst == nullptr)
+    return false;
+
+  // Other incoming node base should match the B base.
+  // SecondInstOffset >= OffsetB && FirstInstOffset > 0?
+  // SecondInstOffset <= OffsetB && FirstInstOffset < 0?
+  // Is non-equal if above are true.
+  APInt FirstInstOffset(Q.DL.getIndexTypeSizeInBits(FirstInst->getType()), 0);
+  FirstInst = FirstInst->stripAndAccumulateConstantOffsets(
+      Q.DL, FirstInstOffset, /* AllowNonInbounds */ true);
```

Comment:
You're skipping over non-inbounds GEPs here, which means that we may wrap around the address space and reach the same address again. I don't think your optimization is valid in the non-inbounds case? (Please add tests.)

Reply:
I don't think so. Shouldn't recursiveGEP_withPtrSub_noninbound2 and recursiveGEP_withPtrSub_noninbound3 fail to fold?

Reply:
Sorry. I looked into a compare test  non_inbounds_gep_compare2 and likely missed the fact that the 2 pointer compares are the same and that is way it should be handled. Maybe a typo in test.
Correction done. Thanks for pointing.

Reply:
is it possible to add alive2 proofs, or does it timeout?

Reply:
The one based on test non_inbounds_gep_compare2 - https://alive2.llvm.org/ce/z/njVxZe
For the recursive case we avoid the folding in case the GEP is non-inbound so it is pretty much the transform - https://alive2.llvm.org/ce/z/FZVjcj